Initial R-2000iB/210F support

demorise opened the pull request:
Adding Support package for r2000ib/210f robot.

demorise committed:
r2000ib_210f support

gavanderhoorn commented:
Thanks for the PR :+1: 

I've restarted Travis as it seemed to have timed-out for some reason.

Let's see what it says.

gavanderhoorn commented:
@ipa-mdl: the Melodic build with the shadow repository seems to be failing (`168.4`).

Is the `industrial_ci` config in this repository too old/incorrect?

mathias-luedtke commented:
@gavanderhoorn: Did you restart it?

gavanderhoorn commented:
Hm, yes, I did.

But I'd already done that a few times, across multiple days.

So I assumed it was going to fail this time again.

But it didn't, so not sure what was going on here.

Thanks for checking @ipa-mdl.

gavanderhoorn committed:
r2000ib: start at version '0.1.0'.

All packages in the experimental repository start at this version.

gavanderhoorn committed:
r2000ib: the package is for all R-2000iB variants.

Not just the /210F (but we do start with it).

gavanderhoorn committed:
r2000ib: be specific about which variant this supports.

gavanderhoorn committed:
r2000ib: mention where specs came from.

gavanderhoorn committed:
r2000ib: update links to tracker and website.

This package sits in fanuc_experimental, so link to that.

gavanderhoorn committed:
r2000ib210f: spec limits in degrees.

And use radians(..) to convert.

Easier to compare with the specifications given by Fanuc.

gavanderhoorn committed:
r2000ib210f: spec link lengths with 4 significant digits.

gavanderhoorn committed:
r2000ib210f: fix mesh normals.

They were inconsistent, so re-orient (and sometimes invert).

gavanderhoorn committed:
r2000ib210f: translate joints in a single dimension only.

This works better with IKFast and some other tools.

Update link_4 mesh to match.

gavanderhoorn committed:
r2000ib210f: correct tool0 orientation.

Link-local frames follow REP-103 (chirality). tool0 matches Fanuc's configuration (ie: Z+ forward).

gavanderhoorn committed:
r2000ib: name roslaunch tests after variant.

Avoid clashes with future additions (ie: other variants).

gavanderhoorn reviewed (commented):
Thanks for the PR @demorise :+1:

Very much appreciated.

I've added a few commits to align things a bit more with other robot support packages.

I've also fixed the normals for all the meshes, as the convex hulls appeared to have inconsistent normals, which won't work reliably with motion planners which try to avoid collisions while planning.

Before:

![rviz_screenshot_2020_07_30-21_29_53](https://user-images.githubusercontent.com/4550046/88969091-9a1eac80-d2b0-11ea-879e-4f31829d7639.png)

After:

![rviz_screenshot_2020_07_30-21_30_07](https://user-images.githubusercontent.com/4550046/88969108-9e4aca00-d2b0-11ea-95aa-005037827d62.png)

(ignore `link_6`: I'd changed orientation of the `link_6` frame (in dd3c2849) before fixing the `collision` mesh orientation for that link.)

Overall I believe this could now be merged.

Seeing as I've changed a few things: could I ask you to verify the model still works for your use-case and is correct?

We'll merge after you give the go-ahead (and CI is green).

gavanderhoorn commented on fanuc_r2000ib_support/package.xml:
```diff
@@ -0,0 +1,67 @@
+    </ul>
+    <p>
+      Joint limits and maximum joint velocities are based on the information in
+      the <em>FANUC R-2000iB Mechanical Unit Operator's Manual</em>
+      version <em>B-82234EN/11</em>. All urdfs are based on the default motion
```
I've added a specific ref to the documentation which provided information on the joint limits et al., as that will make it easier to verify/correct them in the future.

gavanderhoorn commented on fanuc_r2000ib_support/urdf/r2000ib210f_macro.xacro:
```diff
@@ -0,0 +1,182 @@
+    <!-- ROS-Industrial 'tool0' frame: all-zeros tool frame -->
+    <link name="${prefix}tool0" />
+    <joint name="${prefix}link_6-tool0" type="fixed">
+      <origin xyz="0 0 0" rpy="${pi} ${-pi/2.0} 0" />
```
I've changed `tool0`'s orientation to match other Fanuc robots.

@demorise: you ~~have~~ *had* `link_6` rotated `pi` over Y+. Did you do this for a specific reason? We prefer to not diverge from [REP-103](https://www.ros.org/reps/rep-0103.html#coordinate-frame-conventions) for link-local frames (such as `link_6`) and "rectify" the situation with `tool0` or additional toolframes.

demorise replied:
It looks like the CAD file origin of link_6 is oriented different from that of other fanuc robots, so without the rotation, I get something like this when the joint rotates:
![r2000ib_link6_issue](https://user-images.githubusercontent.com/22728175/88975034-2b752b00-d27f-11ea-87dc-a8eaf2d27a02.png)

gavanderhoorn replied:
I'm pretty sure I re-oriented all meshes to follow REP-103, at least for the link-local frames.

I'll check again, but in the meantime: could you make sure you're using the files from this PR?

---

Edit: also: for off-setting mesh orientation/translation, it'd be better to change the `origin` tag of the `visual` and `collision` elements. Not the pose of the `joint` itself.

demorise replied:
I used the files from this PR. My original meshes have a lot of transparent sections.

gavanderhoorn replied:
So this had nothing to do with mesh orientation.

`joint_6` just had an incorrect axis specification. I've corrected it in 5722c8c.

Thanks for spotting that btw, as I'd assumed the axis specs were OK (as it's something which is really easy to notice).

> My original meshes have a lot of transparent sections.

That's the problem with the normals I referenced in 7321ccd.

gavanderhoorn replied:
I'm going to resolve this as I believe I understand why you had that rotation (and why it's no longer needed).

demorise replied:
> That's the problem with the normals I referenced in [7321ccd](https://github.com/ros-industrial/fanuc_experimental/commit/7321ccd3029fa1b59c5c737926280b34fb5f8d51).

Correct. I mentioned that as a confirmation that I was using the correct files. If you don't mind, I'll like to know how you fixed the issue with the normals for future reference.

gavanderhoorn replied:
Meshlab.

Use `filters->Normals, Curvatures and .. ->Re-orient all faces coherently`, then if needed `Invert faces orientation` (from the same menu).

Otherwise I use Blender.

demorise replied:
Thanks. ...and thanks for the effort you've put into this. I just tried the latest commit and it works fine now.
![Screenshot from 2020-07-31 10-00-17](https://user-images.githubusercontent.com/22728175/89048128-ffa48480-d314-11ea-9cff-d3427f1aa45a.png)

gavanderhoorn replied:
Thanks.

Could you take a look at https://github.com/ros-industrial/fanuc_experimental/pull/57#issuecomment-667141786?

gavanderhoorn committed:
r2000ib210f: fix joint_6 axis rotation direction.

Joint rotates around X+.

Checked against Fig 3.1 (Basic Specifications - Robot Configuration) of the operator's manual.

gavanderhoorn commented:
@demorise: if you could compare the output of `rosrun tf2_tools echo.py base tool0` with what the TP shows you that would be the final test I believe we'd need before we can merge this.

Note: `tool0` is an unconfigured/all-zeros tool frame.

demorise commented:
> @demorise: if you could compare the output of `rosrun tf2_tools echo.py base tool0` with what the TP shows you that would be the final test I believe we'd need before we can merge this.
> 
> Note: `tool0` is an unconfigured/all-zeros tool frame.

![IMG_20200731_102653](https://user-images.githubusercontent.com/22728175/89050531-80b14b00-d318-11ea-86ff-9dd1b34876ca.jpg)

![Screenshot from 2020-07-31 10-28-22](https://user-images.githubusercontent.com/22728175/89050667-b5250700-d318-11ea-97d7-56812770f6a3.png)

gavanderhoorn commented:
Nice. Thanks.

(note ROS uses `m` for the translation)

I'll merge this, as CI is green, your checks show expected and correct values and we've also visually verified the model.

gavanderhoorn commented:
I'll squash-merge it, so my additional commits don't end up in the history.

gavanderhoorn commented:
Thanks for the contribution @demorise :+1: 

Let me know if you run into anything you believe is not working as it should when using the package.

Note: because I changed the transform between `joint_3` and `joint_4`, you may need to update your local (detailed) meshes to work with this model.

demorise commented:
Great! Thanks. I'll make the necessary changes.

Beshario commented:
Not sure why the Effort limits are at zero (0) in the URDFs.
The robot model keeps "nodding off" when integrating it to Gazebo!?

gavanderhoorn commented:
The limits are set to `0` because the real limits are unknown.

We don't want to add information to the support packages which cannot be verified or retrieved from authoritative sources.

If you know the actual limits, please contribute them in a PR.

Beshario commented:
Thank you.
